flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/CompoundReducingUpsertWriterState.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.kafka.table;
+
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.table.data.RowData;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+class CompoundReducingUpsertWriterState<WriterState> {
```

Comment:
I'd drop the compound here. Imho many writers will have compounds states and I wouldn't want to make it the default.

Reply:
On a second thought, we could have a generic Compound state with a generic writer. Then the name would be very fitting.

Reply:
I am in favor of your first comment because afaik there will be some efforts to integrate the `ReducingUpsert` functionality in the table API operators and make it available to all upsert sinks.
Therefore generalizing it now may not have an impact because it is dropped soonish.

Reply:
I'm okay with that. I just think that it could be a general building block for advanced sources/sinks. But we can generalize it then.